lblod/ember-rdfa-editor
# Improved RDFa backwards compatibility

### Overview
This PR attempts to improve the backwards compatibility of rdfa-aware nodes.
Specifically, it converts each rdfa-aware nodespec to a function producing a node-spec based on an options argument, which is defined as follows:
```typescript
type Config = {
  rdfaAware?: boolean;
};
```
This, of course includes a breaking change as a lot of the nodespecs have been converted to functions. Specifically, the following node-specs have been converted to a function producing a node-spec:
- `block_rdfa`
- `inline_rdfa`
- `invisible_rdfa`
- `repaired_block`
- `list_item`, `bullet_list` and `ordered_list`
- `heading`

The following mark-specs have also been converted to a function:
- `inline_rdfa`

The following node-specs where already produced by a function (their options argument has been extended to include the `rdfaAware` option)
- the table node-specs
- `link`

The `doc` node-spec has been dropped and the `docWithConfig` function has been renamed to `doc`

The `rdfaAware` option always has a default value of false, the new features are thus opt-in.

Regarding the dummy pages:
- The 'index', 'dummy plugins' and 'space invisible' pages are not rdfa-aware
- The `editable nodes POC` page is rdfa-aware

### How to test/reproduce
- Start the dummy app

- Visit the `index`,  'dummy plugins' or 'space invisible' page
- Notice, (using e.g. the prosemirror devtools) that the different node-specs use the classic RDFa attributes
- Using the different sample configs, notice that the RDFa stays preserved (using the classic method)

- Visit the `editable nodes POC` page
- Notice that the different RDFa node-specs use the new API
- Using the different sample configs, notice that the RDFa stays preserved (using the new API)
- The 'editable' RDFa nodes should work as expected

### Challenges/uncertainties
This PR contains quite a few breaking changes, as it converts some node-specs to functions producing node-specs.
**Initially I tried the following approach:**
- Define an optional `rdfaAware` boolean property on node-specs, this property can be supplied to a node-spec using the spread operator: e.g. `block_rdfa: {...block_rdfa, rdfaAware: true }`
- Adapt the `toDOM`, `parseDOM` and `attrs` properties of a node-spec to be reactive to this `rdfaAware` property

This approach had the big following advantage:
- No (new) breaking changes: the `rdfaAware` property has a default value of false/undefined. The node-specs stays node-specs and do not become functions

But this approach did not work out:
- The `toDOM` and `parseDOM` node-spec properties get pulled out of their contexts, so it is not possible to get access to the `this` object inside them. I did not (yet) see an other option on how to access `this.rdfaAware`.

**In order to limit the breaking changes, we could also take the following approach:**
- Introduce seperate `configurable` functions that produce node-specs:
  * `configurableBlockRDFa`
  * `configurableInlineRDFa`
  * `configurableList`
  * ...
- Keep the current `block_rdfa`, `inline_rdfa` ... variables as normal node-specs with `rdfaAware: false`

This approach does have the disadvantage to introduce quite a few new variables/functions.

### Checks PR readiness
